api/service/discovery/service.go

```diff
@@ -74,19 +79,21 @@ func (s *Service) Init() {
 	log.Info("Init discovery service")
 }
 
-func (s *Service) pingPeer(peer p2p.Peer) {
+func (s *Service) pingPeer() {
+	tick := time.NewTicker(5 * time.Second)
 	ping := proto_discovery.NewPingMessage(s.host.GetSelfPeer())
 	buffer := ping.ConstructPingMessage()
 	content := host.ConstructP2pMessage(byte(0), buffer)
+
+	for {
```

Comment:
Infinite loop (goroutine leak)?  We need to exit the loop on something….  How about until we receive a matching pong?

Reply:
periodically ping is to make sure the public key of consensus is received by everyone in the shard.  It's not a bad idea to stop sending until we received a pong.  I'll add it in later PR.